### PR TITLE
Updated #5829: Added --editor (-e) option to open all generated & copied files in the user's editor

### PR DESCRIPTION
**Pull request by ndbroadbent**
This is an updated version of #5829, with [@rafaelfranca's proposal to make it higher level](https://github.com/rails/rails/pull/5829#issuecomment-7098334).

The `--editor` option is now aliased as `-e`, and is available for all generators that inherit from `Rails::Generators::Base`. The default editor is now `GUI_EDITOR`, and falls back to `EDITOR`.

It wraps Thor's `template` and `copy_file` actions, and opens generated/copied files in your text editor if `options["editor"]` is present.

P.S. I initially tried to just open the main generated files, but I think it's better to just open them all. If you run the model generator, you'll probably want to edit the migration, model, and tests. (Instead of just the model.)

### CODE REVIEW

**Comment by rafaelfranca** [comment]
@matthewd not sure if I got you suggestion. Are you saying to extract the open in editor code to a new method or are you thinking in a more generic handler for the main file?

**Comment by matthewd** [comment]
@rafaelfranca the latter. Nothing complex... but just some agnostic way of knowing which file is the interesting one. Maybe as simple as 

```ruby
def create_mailer_file
  destination = File.join('app/mailers', class_path, "#{file_name}.rb")
  template "mailer.rb", destination
  primary_file destination
end

[..]
def primary_file(filename)
  # I'm not sure exactly what, but I can see other things wanting to hook in here

  open_file_in_editor(filename) if options["editor"].present?
end
```

Largely, it seems neater to keep the raw "open in editor" concept out of the individual generators.

I'm not sure how, but the logic in https://github.com/rails/rails/pull/8553/files#r5010513 could hopefully live there too. Maybe?

File: railties/lib/rails/generators/actions.rb [hunk 1]
```diff
@@ -232,6 +232,16 @@ def readme(path)
         log File.read(find_in_source_paths(path))
       end
 
+      # Opens the given file in the user's editor, if generator is run with the <tt>--editor</tt> option.
+      #
+      #   open_file_in_editor "app/models/account.rb"
+      def open_file_in_editor(path)
+        # Attempt to only open the main file generated by the base generator.
+        if shell.base.class === self || shell.base.class.invocations.key?(:orm)
+          run("#{options["editor"]} \"#{path}\"")
```
Comment (repinel) [inline railties/lib/rails/generators/actions.rb]: Have you tested with command-line editor like `vim`? Although it opens template, the generator is most likely still running and also outputting to the terminal.

**Comment by sgrif** [comment]
Forcing a rebuild.

**Comment by sgrif** [comment]
Ok well I can't get travis to run it, but I pulled this branch down locally and it has failing tests. @ndbroadbent are you still interested in working on this?

**Comment by sgrif** [comment]
Closing due to inactivity.

**Pull request closed without merging**